File: backend/app/schemas/metrics.py
import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.models.site_status import HealthStatus

MetricName = Literal["carbon", "biodiversity", "ndvi"]


class MetricPointOut(BaseModel):
    model_config = {"from_attributes": True}

    recorded_at: date
    carbon_tco2e: float
    biodiversity_index: float
    ndvi: float


class MetricsOut(BaseModel):
    site_id: uuid.UUID
    range_days: int
    points: list[MetricPointOut]


class ForecastPointOut(BaseModel):
    recorded_at: date
    value: float
    lower: float
    upper: float


class ForecastOut(BaseModel):
    site_id: uuid.UUID
    metric: MetricName
    historical: list[ForecastPointOut]
    forecast: list[ForecastPointOut]
    model: str
    confidence: float = Field(ge=0, le=1)


class StatusOut(BaseModel):
    site_id: uuid.UUID
    status: HealthStatus
    anomaly_score: float
    last_evaluated_at: datetime
    reasons: list[str]


class SiteMetricSummary(BaseModel):
    model_config = {"from_attributes": True}

    site_id: uuid.UUID
    site_name: str
    health: HealthStatus
    anomaly_score: float
    latest: MetricPointOut
    changes: dict[str, float]


class ProjectMetricsOut(BaseModel):
    project_id: uuid.UUID
    sites: list[SiteMetricSummary]
    portfolio_points: list[MetricPointOut]

File: frontend/src/types/metrics.ts
export type MetricName = "carbon" | "biodiversity" | "ndvi";
export interface SiteMetricPoint { recorded_at: string; carbon_tco2e: number; biodiversity_index: number; ndvi: number; }
export interface MetricsResponse { site_id: string; range_days: number; points: SiteMetricPoint[]; }
export interface ForecastPoint { recorded_at: string; value: number; lower: number; upper: number; }
export interface ForecastResponse { site_id: string; metric: MetricName; historical: ForecastPoint[]; forecast: ForecastPoint[]; model: string; confidence: number; }
export interface StatusResponse { site_id: string; status: "healthy" | "watch" | "at_risk"; anomaly_score: number; last_evaluated_at: string; reasons: string[]; }
export interface ProjectMetricSummary { site_id: string; site_name: string; health: "healthy" | "watch" | "at_risk"; anomaly_score: number; latest: SiteMetricPoint; changes: { carbon: number; biodiversity: number; ndvi: number }; }
export interface ProjectMetricsResponse { project_id: string; sites: ProjectMetricSummary[]; portfolio_points: SiteMetricPoint[]; }
